pkg/config/remote/service/service_test.go

```diff
@@ -84,6 +86,93 @@ func newTestService(t *testing.T, api *mockAPI, uptane *mockUptane, clock clock.
 	return service
 }
 
+func TestServiceBackoff(t *testing.T) {
+	api := &mockAPI{}
+	uptaneClient := &mockUptane{}
+	clock := clock.NewMock()
+	service := newTestService(t, api, uptaneClient, clock)
+
+	lastConfigResponse := &pbgo.LatestConfigsResponse{
+		TargetFiles: []*pbgo.File{{Path: "test"}},
+	}
+
+	api.On("Fetch", mock.Anything, &pbgo.LatestConfigsRequest{
+		Hostname:                     service.hostname,
+		AgentVersion:                 version.AgentVersion,
+		CurrentConfigSnapshotVersion: 0,
+		CurrentConfigRootVersion:     0,
+		CurrentDirectorRootVersion:   0,
+		Products:                     []string{},
+		NewProducts:                  []string{},
+	}).Return(lastConfigResponse, errors.New("simulated HTTP error"))
+	uptaneClient.On("State").Return(uptane.State{}, nil)
+	uptaneClient.On("Update", lastConfigResponse).Return(nil)
+
+	// We'll set the default interal to 1 second to make math less hard
+	service.defaultRefreshInterval = 1 * time.Second
+
+	// There should be no errors at the start
+	assert.Equal(t, service.backoffErrorCount, 0)
+
+	err := service.refresh()
+	assert.NotNil(t, err)
+
+	// We should be tracking an error now. With the default backoff config, our refresh interval
+	// should be somewhere in the range of 1 + [30,60], so [31,61]
+	assert.Equal(t, service.backoffErrorCount, 1)
+	refreshInterval := service.calculateRefreshInterval()
+	assert.GreaterOrEqual(t, refreshInterval, 31*time.Second)
+	assert.LessOrEqual(t, refreshInterval, 61*time.Second)
+
+	err = service.refresh()
+	assert.NotNil(t, err)
+
+	// Now we're looking at  1 + [60, 120], so [61,121]
+	assert.Equal(t, service.backoffErrorCount, 2)
+	refreshInterval = service.calculateRefreshInterval()
+	assert.GreaterOrEqual(t, refreshInterval, 61*time.Second)
+	assert.LessOrEqual(t, refreshInterval, 121*time.Second)
+
+	err = service.refresh()
+	assert.NotNil(t, err)
+
+	// After one more we're looking at  1 + [120, 240], so [121,241]
+	assert.Equal(t, service.backoffErrorCount, 3)
+	refreshInterval = service.calculateRefreshInterval()
+	assert.GreaterOrEqual(t, refreshInterval, 121*time.Second)
+	assert.LessOrEqual(t, refreshInterval, 241*time.Second)
+
+	// Make the call succeed so we can test the recovery interval
+	api = &mockAPI{}
+	api.On("Fetch", mock.Anything, &pbgo.LatestConfigsRequest{
+		Hostname:                     service.hostname,
+		AgentVersion:                 version.AgentVersion,
+		CurrentConfigSnapshotVersion: 0,
+		CurrentConfigRootVersion:     0,
+		CurrentDirectorRootVersion:   0,
+		Products:                     []string{},
+		NewProducts:                  []string{},
+	}).Return(lastConfigResponse, nil)
+	service.api = api
```

Comment:
This seems like a second test case. Could we split this test into more "unit" like one ?

Reply:
Yep, can do!